Nested Combobox Grouping

awseward opened the pull request:
This PR enables the `Combobox` to group items in an n-tiered fashion.

**Note**: Please ignore all changes made where are not under the `src/` directory.
### Before

<img width="514" alt="screen shot 2016-04-19 at 15 32 14" src="https://cloud.githubusercontent.com/assets/901035/14657631/f2253e3c-0643-11e6-849b-0a65e449b4d0.png">
### After

<img width="511" alt="screen shot 2016-04-19 at 15 28 02" src="https://cloud.githubusercontent.com/assets/901035/14657544/690df06c-0643-11e6-8a45-ec63e8347346.png">

awseward committed:
[TEMPORARY] Ignore dist/, docs/, lib/

awseward committed:
Build hierarchical object

Next step is to read it in the dropdown

awseward committed:
Keep track of paths (much like keys)

awseward committed:
Check to see if paths are being set in the state properly

awseward committed:
Decided on using a tree of some sort

awseward committed:
Start adding _orderedKeys to _setIn objects

awseward committed:
Add _orderedKeys to objects from _setIn

awseward committed:
[WIP] Try out flattening to an items array

awseward committed:
Maybe this will work, but not too sure, TBH

awseward committed:
Remove a bunch of console warns, other cleanup

Starting to wonder if this isn't just going to be just a general pain to
implement without breaking away pretty significantly from what jquense
has going on in here.

awseward committed:
Getting closer

Now getting errors with same keys

awseward committed:
Getting yet closer.

Now need to make sure the headers have unique keys...

awseward committed:
Make group header keys unique again

awseward committed:
Distinguish path for keys from label

awseward committed:
It more or less works!

awseward committed:
Add ListMultiGroupable

Will simplify things to just have it be a separate file, although I
think eventually the distinction between ListGroupable and
ListMultiGroupable will no longer even matter at all... One would think
that a ListGroupable will really just be a single-level
ListMultiGroupable.

awseward committed:
Use new groupable in Combobox when appropriate

awseward committed:
Return ListGroupable to its original state

Best to just leave that alone...

awseward committed:
Remove ListGroupable code behind short circuit

ListMultiGroupable can assume that what it's being given is to be
grouped in a nested fashion.

awseward committed:
Comment out console warns

awseward committed:
Add _data() back into ListMultGroupable

TODO: Actually implement it. It's just hardcoded to [] for now.

awseward committed:
Print result of _data() in ListGroupable

awseward committed:
Some renaming/cleanup

awseward committed:
Implement _data() in ListMultiGroupable

At least, I think so. Won't really know until I try it out here

awseward committed:
Remove some _data() console.warn calls

awseward committed:
Uncomment move, chip away at getItemDOMNode

This thing is funky...

awseward committed:
Put some sane formatting into ListMultiGroupable

awseward committed:
Print ListGroupable getItemDOMNode list var

awseward committed:
Log ListMultiGroupable getItemDOMNode list

awseward committed:
[REMOVE] Add a stripped down build script

awseward committed:
Need to render items w/ an aggregate offset

This is a first attempt at that, except the offset never gets actually
incremented past 0 yet, so the net impact of this should be nothing for
now...

awseward committed:
Implement aggregate offset

awseward committed:
Return offset from the function itself

awseward committed:
More accurate name for this god awful function

awseward committed:
Remove console warn

awseward committed:
I think this might just work...

awseward committed:
Getting closer

Not quite offsetting correctly still, though.

Added some logging for diagnostic purposes.

awseward committed:
Add offset to foundIndex

Otherwise, there's no point to it at all!

awseward committed:
Not sure, but maybe check found in reduce call

awseward committed:
Finish getItemDOMNode

Finally works consistently.

Also a bunch of clean up in this commit... (Not too stoked about that,
as it distracts from the actual change made in the commit, but oh well)

awseward committed:
[Tentative] Remove unneeded code

awseward committed:
Some cleanup

Get the formatting/style to a little bit more of a sane state...

awseward committed:
Add depth to optgroup, option

awseward committed:
[Tentative] Clean up the monster render stuff

Add depth...

awseward committed:
Fix depth stuff

Works properly now

awseward committed:
Add depth styles

awseward committed:
Clean up the sass a little bit

It's still not being served properly, but that can be figured out later.

Also, this could probably use some better backwards compatibility.
Should avoid stepping on existing styles a little better...

awseward committed:
Pull out GroupHeader component

awseward committed:
First iteration of object traversal helper

It's almost there...

The path stuff isn't quite right, though

awseward committed:
Fix the bugs in the objectTraversal util

awseward committed:
Clean some things up in objectTraversal

awseward committed:
[WIP] Try objectTraversal in ListMultiGroupable

awseward committed:
Fix offsets (almost)

awseward committed:
Remove unnecessary comment

awseward committed:
Clean out some cruft

awseward committed:
Revert .gitignore to @master

awseward committed:
Add dist/, docs/, lib/ back in

yoyis3000 commented on src/GroupHeader.jsx:
```diff
@@ -0,0 +1,29 @@
+import CustomPropTypes from './util/propTypes';
+
+function GroupHeader(props) {
+  const {
```
NB: Why not get className and id while you're at it?

awseward replied:
@yoyis3000 I usually try to hold off on destructuring things until I have to reference them more than once in the enclosing scope.

kespinola commented on src/Combobox.jsx:
```diff
@@ -148,7 +155,7 @@ var ComboBox = React.createClass({
       , listComponent: List } = this.props;
 
-    List = List || (groupBy && GroupableList) || PlainList
+    List = List || _resolveListGroupable(groupBy) || PlainList
```
`DropdownList` and `Combobox` take a `listComponent` prop. What do you think of exposing `MultiGroupList` from the package then passing `MultiGroupList` to `listCompoent` when you instaniate a new `Combobox` like so

```
import { MultiGroupList, Combobox } from 'react-widgets';

render() {
  <Combobox listComponent={MultGroupList} />
}
```

Naturally extending the functionality using the given api instead of adding checks for our special case.

awseward replied:
Oh I like that a lot. Awesome, thanks!

yoyis3000 commented:
+1

awseward committed:
Pass ListMultiGroupable into Combobox

A much more open/closed approach.

Great suggestion by @ksespinola here:
https://github.com/procore/react-widgets/pull/12#discussion_r60326687

awseward committed:
Update built artifacts, docs

awseward committed:
Sort module exports again

I added ListMultiGroupable in at the end and forgot to sort.

kespinola commented on src/ListMultiGroupable.jsx:
```diff
@@ -0,0 +1,477 @@
+    // Conveniently, someArray[-1] gives undefined, which is just what we want
+    return list[index];
+  }
+});
```
Mind giving me a run down of this when you get in? Thanks.

awseward replied:
I'm in now, so anytime before 9:30 is good for me.

farism commented on lib/ListGroupable.js:
```diff
@@ -236,8 +236,7 @@ exports['default'] = _react2['default'].createClass({
     var groups = this.state.groups;
-
-    return this.state.sortedKeys.reduce(function (flat, grp) {
+    return result = this.state.sortedKeys.reduce(function (flat, grp) {
```
n/b: Unnecessary whitespace diff

farism commented on lib/index.js:
```diff
@@ -11,12 +11,13 @@ if (process.env.NODE_ENV !== 'production') {
+  DropdownList: require('./DropdownList'),
+  ListMultiGroupable: require('./ListMultiGroupable'),
   Multiselect: require('./Multiselect'),
+  NumberPicker: require('./NumberPicker'),
```
n/b: Same as above w/ whitespace, if we are going to hack on 3rd party modules we should only do the minimum required changes.

farism commented on lib/less/core.less:
```diff
@@ -211,15 +211,34 @@ ul.rw-list {
+        &-5 {
+          padding-left: 60px;
+        }
+      }
```
I think you're looking for the `@for` control directive here

```
@for $i from 0 through 5
  &-#{$i}
    width: ($i * 10) + 10px
```

or something like that.

farism commented on lib/less/core.less:
```diff
@@ -211,15 +211,34 @@ ul.rw-list {
-      padding-left: 20px;
+    > li {
+      &.rw-list-depth {
+        // TODO: There's probably some kind of expression syntax for this...
```
I think you're looking for the `@for` control directive here

```
@for $i from 0 through 5
  &-#{$i}
    width: ($i * 10) + 10px
```

or something like that.

farism commented on src/less/core.less:
```diff
@@ -211,15 +211,34 @@ ul.rw-list {
-      padding-left: 20px;
+    > li {
+      &.rw-list-depth {
+        // TODO: There's probably some kind of expression syntax for this...
```
loops in LESS are kind of gross

farism commented on src/index.js:
```diff
@@ -13,13 +13,14 @@ if (process.env.NODE_ENV !== 'production' ) {
+  ListMultiGroupable:  require('./ListMultiGroupable'),
+  Multiselect:         require('./Multiselect'),
+  NumberPicker:        require('./NumberPicker'),
+  SelectList:          require('./SelectList'),
```
n/b: Same as above w/ whitespace, if we are going to hack on 3rd party modules we should only do the minimum required changes.

farism commented on src/ListGroupable.jsx:
```diff
@@ -210,9 +210,8 @@ export default React.createClass({
-
-    return this.state.sortedKeys
-      .reduce( (flat, grp) => flat.concat(groups[grp]), [])
+    return result = this.state.sortedKeys
```
n/b: Unnecessary whitespace diff

farism commented on package.json:
```diff
@@ -45,6 +45,7 @@
     "build:docs": "npm run clean:docs && npm run compile:docs && npm run assets:docs",
     "build:dist": "npm run clean:dist && npm run compile:dist & npm run compile:locale && npm run less && npm run assets:dist",
     "build": "npm run build:lib && npm run build:dist & npm run build:docs",
+    "b": "npm run build:lib && npm run build:dist",
```
Do we need this shortcut run script in our fork?

awseward replied:
@farism Good catch, forgot to remove this.

awseward committed:
Reduce diff footprint

awseward committed:
Undo temporary changes to package.json

awseward committed:
Clean up path stuff

awseward committed:
Abort getItemDOMNode if no input

No need to try and find an item's DOM node if that item is
undefined.

awseward committed:
Clean up comments

remyyounes commented on src/ListMultiGroupable.jsx:
```diff
@@ -0,0 +1,477 @@
+  shouldPushKey && obj._orderedKeys.push(key);
+}
+
+function _setIn(obj, path, val) {
```
Switching your arguments to have obj always last would allow for FP.

awseward replied:
Ooh I like. May do this in a followup

kespinola commented:
+1

remyyounes commented on lib/ListMultiGroupable.js:
```diff
@@ -0,0 +1,468 @@
+  return outputArray;
+}
+
+function _renderHeadersAndItems(groupedData, processHeader, processItems) {
```
where are these being called ?

awseward replied:
Some of these are old functions I forgot to pull out.

Thanks for catching them

awseward committed:
Remove some unused left-over code

remyyounes commented:
+1

thomasm-procore commented:
Passed testing.
